File: Sort/heap-sort.js
function heapify(array, n, i){
    let largest = i;
    let l = 2 * i + 1;
    let r = 2 * i + 2;
    if(l < n && array[i] < array[l]){
        largest = l;
    }
    if(r < n && array[largest] < array[r]){
        largest = r;
    }
    if(largest != i){
        array[i] = array[i] + array[largest];
        array[largest] = array[i] - array[largest];
        array[i] = array[i] - array[largest];
        heapify(array, n, largest);
    }
}

function sort(array){
    for(let i = Math.floor(array.length / 2); i > -1; i --){
        heapify(array, array.length, i);
    }
    for(let j = array.length - 1; j > 0; j --){
        array[j] = array[j] + array[0];
        array[0] = array[j] - array[0];
        array[j] = array[j] - array[0];
        heapify(array, j, 0);
    }
}

let a = [5, 6, 2, 9, 1];

sort(a);

console.log(a);

File: Sort/heap-sort.go
package main

import (
	"bufio"
	"fmt"
	"math"
	"os"
	"strconv"
	"strings"
)

func Heapify(array []int, n int, i int) {
	largest := i
	l := 2*i + 1
	r := 2*i + 2
	if l < n && array[i] < array[l] {
		largest = l
	}
	if r < n && array[largest] < array[r] {
		largest = r
	}
	if largest != i {
		array[i], array[largest] = array[largest], array[i]
		Heapify(array, n, largest)
	}
}

func Sort(array []int) {
	for i := int(math.Floor(float64(len(array) / 2))); i > -1; i-- {
		Heapify(array, len(array), i)
	}
	for j := len(array) - 1; j > 0; j-- {
		array[j], array[0] = array[0], array[j]
		Heapify(array, j, 0)
	}
}

func main() {
	scanner := bufio.NewScanner(os.Stdin)
	scanner.Scan()
	s := scanner.Text()
	var a []int
	for _, value := range strings.Split(s, " ") {
		x, _ := strconv.Atoi(value)
		a = append(a, x)
	}
	Sort(a)
	fmt.Println(a)
}
